### PR TITLE
add vjp interface for reshard op.

### PR DESCRIPTION
**Pull request by winter-wang**
### PR Category

Auto Parallel

### PR Types

New features

### Description

- 规范ReshardOp的名字。 （从ReShardOp调整为ReshardOp）
- 完善reshard的python api， 支持pir模式以及传递placement参数。
- reshard增加vjp接口，支持带reshard op的网络跑自动微分流程。

### Other
pcard-67164

### CODE REVIEW

**Review by wanghuancoder — DISMISSED** [review]
LGTM

File: python/paddle/distributed/auto_parallel/api.py [hunk 1]
```diff
@@ -388,6 +392,9 @@ def reshard(dist_tensor, mesh, placements):
             dist_attr._set_partial_dims(partial_dims)
 
         return paddle.base.core.reshard(dist_tensor, dist_attr)
+    elif in_pir_mode():
+        dim_map = to_dim_map(placements, dist_tensor.ndim)
+        return paddle._pir_ops.reshard(dist_tensor, mesh, dim_map, placements)
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py]: The api should take dims_mapping or placement, but not both.

File: test/auto_parallel/pir/test_reshard.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                    self.fwd_input = op.operand_source(0)
+                    self.assertEqual(
+                        self.fwd_input.dist_attr().dims_mapping, [-1, -1]
+                    )
+                    self.assertEqual(
+                        self.fwd_input.dist_attr().partial_dims, set()
+                    )
+                    self.assertEqual(
+                        self.fwd_input._local_shape,
+                        [BATCH_SIZE, IMAGE_SIZE // 2],
```
Comment (JZ-LIANG) [inline test/auto_parallel/pir/test_reshard.py]: reshard input is replicated, so its local shape should be equal to global shape == [BATCH_SIZE, CLASS_NUM]
"IMAGE_SIZE // 2" is ambiguous

File: python/paddle/distributed/auto_parallel/api.py (hunk 1 above)
Comment (winter-wang) [reply python/paddle/distributed/auto_parallel/api.py]: done, thanks!

**Review by zhiqiu — DISMISSED** [review]
LGTM

File: test/auto_parallel/pir/test_reshard.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                    self.assertEqual(
+                        self.fwd_output.dist_attr().partial_dims, set()
+                    )
+                    self.assertEqual(
+                        self.fwd_output._local_shape,
+                        [BATCH_SIZE / 2, IMAGE_SIZE // 2],
+                    )
+                elif index == 1:
+                    # backward reshard op
+                    self.assertEqual(op.result(0).type(), self.fwd_input.type())
```
Comment (JZ-LIANG) [inline test/auto_parallel/pir/test_reshard.py:95]: should also check the input of backward reshard 
type could not be equal, dist attr should be equal

File: test/auto_parallel/pir/test_reshard.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                index += 1
```
Comment (JZ-LIANG) [inline test/auto_parallel/pir/test_reshard.py:96]: should check the position of two reshard op in program.

File: test/auto_parallel/pir/test_reshard.py (hunk 3 above)
Comment (winter-wang) [reply test/auto_parallel/pir/test_reshard.py:95]: ok

File: test/auto_parallel/pir/test_reshard.py (hunk 2 above)
Comment (winter-wang) [reply test/auto_parallel/pir/test_reshard.py]: done ,thanks!

File: test/auto_parallel/pir/test_reshard.py (hunk 4 above)
Comment (winter-wang) [reply test/auto_parallel/pir/test_reshard.py:96]: ok!

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Pull request merged**